### ISSUE TITLE
Measuring "tentative installations"

### ISSUE THREAD

**Issue by maroneze** [ocaml/opam#3284]
Is there a way to measure or estimate somehow the number of "tentative installations" of a given package?

In practice, the situation is the following: given our package A depends on specific versions of B, if the user has a newer/exotic version of B, it will be unable to install A (or will have to downgrade). It costs us time to maintain compatibility of A with multiple versions of B, so unless users ask explicitly, we can't do it for *every* version of B.

However, new potential users who already have B and would like to install A may just give up after the errors, so we will never hear from them. But it would be useful to know how many users *tried* to install our package, but ultimately failed to do so (or cancelled, or something else; the exact reason is not important).

I believe OPAM does not track anything related to that, but is there something vaguely close? For instance, one could count the number of accesses to the `.tar.gz` archive for package A, but since the download only happens *after* constraint resolution, in my example the user would never even try downloading it at all, so we will not know about them.

**Comment by AltGr** [comment]
Indeed, that would be useful. At the moment, we don't track anything, except hits to the cache of archives on opam.ocaml.org which are logged by the web server (and used to display the package download stats on the site, which aren't all that useful considering `ocamlfind` is guaranteed first place...)

We have considered gathering stats on package installations (with user consent), and the ["solver farm"](https://github.com/IRILL/cudf_remote_proxy/blob/master/cudf_remote_proxy) that could at some point be used to remotely resolve package installations could have been a means to do that (besides providing a large test basis for improving the solvers). But we didn't get the time to do it, and the service has now been down for a while.

We could of course add a simple tracker of opam requests/responses, but despite its usefulness, a mix of lack of interest and privacy concerns have stalled such attempts until now. Note that your case is a bit more complex, and would need some extra info to be inferred from request+response+"user said yes/no": opam will provide a solution to install A no matter what, up/downgrading B if necessary, and wether this was acceptable or not can only be inferred from the user's response.

But at least that would provide more meaningful stats on opam.ocaml.org.

**Comment by maroneze** [comment]
Ok, thanks for the information, I was kind of expecting that.

As far as I understand, currently the contents of the `messages` field in an opam package is only displayed *after* constraint resolution (i.e. just before installing the package, but when we are sure we can do it), right?

If so, I wonder if it would make sense to adapt the current `messages` field to include some kind of pre-installation message, or to add a sort of `pre-messages` field, that would be displayed *before* constraint resolution, and which could contain something such as "If you have trouble installing this, click here to tell us: https://mydomain.com/opam-installation-error-form".

Two issues I already see with my suggestion are:

1. It would still be necessary to retrieve information about the users' opam configuration, so in the end it will be not much simpler than a direct bug report. But it could somehow be a bridge between mobile app stores, which lack privacy, but where user feedback only requires typing the message + 2 clicks, and traditional package managers, which are devoid of feedback mechanisms other than users going all the way to submit bug reports.

2. There is the risk that packages will become too chatty/spammy, so maybe a `fail-messages` field would be more like it, as in messages emitted only when installing the package fails. But I have no idea if this actually makes sense for users and package developers (my experience is limited to developing very few packages).
